shared/src/main/scala/mlscript/Typer.scala

```diff
@@ -145,8 +159,19 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool) extend
       case _ => Set.empty // TODO TupleType?
     }
   
+  def targsMapOf(ty: Type)(implicit ctx: Ctx, raise: Raise, lookup: Option[(Str, Option[Str])] = N,
+      vars: Map[Str, SimpleType] = Map.empty): Map[Str, SimpleType] = ty match {
+    case Inter(l, r) => targsMapOf(l) ++ targsMapOf(r)
+    case AppliedType(b, ts) => ctx.tyDefs.get(b.name).fold(Map.empty[Str, SimpleType]) {
+      td => td.tparams.map(_.name).zip(ts.map(typeType(_)(ctx.nextLevel, raise, lookup, vars))).toMap
+    }
+    case _ => Map.empty
```

Comment:
This function extracts the type arguments applied to the base class from the class body. Classes can only inherit from a conjunction of `Primitive`s, `AppliedType`s and `Record`s, and the only ones with applied type arguments are `AppliedType`s. I guess it wasn't clear that this is only supposed to be used on the body of a type definition.

The recursive function is now wrapped in a function taking a `TypeDef`, and the legal cases where an empty map should be returned are now listed.

Reply:
Okay, but that doesn't work, even without traits, because types need to be inspected semantically, not syntactically:

```ts
// FIXME should raise error
:e
class Class3A[A]
    method MtdA: A
type Type3A = Class3A[string]
class Class3B: Type3A
    method MtdA = 1
//│ Defined class Class3A
//│ Declared Class3A.MtdA: class3A -> nothing
//│ Defined type Type3A
//│ Defined class Class3B
//│ Defined Class3B.MtdA: nothing -> 1
```
(On an unrelated note, the `Class3B.MtdA: nothing -> 1` is strange.)

The _source of truth_ for figuring out class parents is the `checkParents` function, which is the one that checks what is legal or illegal inheritance. It currently only returns a boolean because at the time we did not ned anything else, but if we need more info, we should get it from there. It's better than duplicating the parents-analysis logic and making non-local assumptions that are not obvious and may thus break in the future.

Reply:
On the other hand, `checkParents` does not traverse inherited parents. So I feel like extending `checkCycleComputeFields` to also glean method declarations and definitions from parents (both classes and traits) would be the better way to go. This is a proper tree traversal which, which is more appropriate for detecting problems like this: https://github.com/hkust-taco/mlscript/pull/22/commits/a1c3b47d22ee3fc62332250c5cfa96e3f1a3e694#diff-271f81eabb0cf04b1f9fe33d0620031626c2c2cb6c311ebafa625a54caf951cbR268

I think in the extended `checkCycleComputeFields` traversal, we could in a single go:

- accumulate a set of method declaration signatures from parent types, transitively, using the proper shadowing semantics; there might be several signatures coming from several class and trait parents, and the method definition will have to satisfy them all

- accumulate method definitions that do not have a signature, if there are any – these  will be used to block further overriding of the method

- compute the type arguments mapping on the fly: when traversing a typedef, we pass a map that we will use to `typeType` the method signatures we find, so that they are immediately instantiated at the right types

I understand this is quite a refactoring, and I actually think we should do this in a later PR, after merging this one, which has been pending long enough (I need to merge it to unblock some of my own changes, too).